R/exercise.R

```diff
@@ -918,21 +918,108 @@ exercise_check_code_is_parsable <- function(exercise) {
+  default_message <- i18n_span(
+    "text.unparsable",
+    HTML(i18n_translations()$en$translation$text$unparsable)
   )
+  unicode_message <- exercise_check_unparsable_unicode(exercise)
+
+  feedback <- list(
+    message = HTML(unicode_message %||% default_message),
+    correct = FALSE,
+    location = "append",
+    type = "error"
+  )
+
+  exercise_result_error(error$message, feedback)
+}
+
+exercise_check_unparsable_unicode <- function(exercise) {
+  code <- exercise[["code"]]
+
+  # Early exit if code is made up of all ASCII characters ----------------------
+  if (!grepl("[^\\x00-\\x7F]", code, perl = TRUE)) {
+    return(NULL)
+  }
+
+  # Check if code contains Unicode quotation marks -----------------------------
+  single_quote_pattern <- "[\u2018\u2019\u201A\u201B\u275B\u275C\uFF07]"
+  double_quote_pattern <- "[\u201C-\u201F\u275D\u275E\u301D-\u301F\uFF02]"
+  quote_pattern <- paste(single_quote_pattern, double_quote_pattern, sep = "|")
+
+  if (grepl(quote_pattern, code)) {
+    character <- str_extract(code, quote_pattern)
+    lint <- exercise_highlight_unparsable_unicode(code, quote_pattern)
+
+    # Replace curly single quotes with straight single quotes
+    suggestion <- gsub(single_quote_pattern, "'", code)
+    # Replace all other Unicode quotes with straight double quotes
+    suggestion <- gsub(double_quote_pattern, '"', suggestion)
+    suggestion <- html_code_block(suggestion)
+
+    return(
+      build_unparsable_i18n_span(
+        "unparsablequotes", character, lint, suggestion
+      )
+    )
+  }
+
+  # Check if code contains Unicode dashes --------------------------------------
+  # Regex searches for all characters in Unicode's general category
+  #   "Dash_Punctuation", except for the ASCII hyphen-minus
+  #   (https://www.unicode.org/reports/tr44/#General_Category_Values)
+  dash_pattern <- paste0(
+    "[\u00af\u05be\u06d4\u1400\u1428\u1806\u1b78\u2010-\u2015\u203e\u2043",
+    "\u2212\u23af\u23e4\u2500\u2796\u2e3a\u2e3b\u30fc\ufe58\ufe63\uff0d]"
+  )
+
+  if (grepl(dash_pattern, code)) {
+    character <- str_extract(code, dash_pattern)
+    lint <- exercise_highlight_unparsable_unicode(code, dash_pattern)
+
+    # Replace Unicode dashes with ASCII hyphen-minus
+    suggestion <- gsub(dash_pattern, "-", code)
+    suggestion <- html_code_block(suggestion)
+
+    return(
+      build_unparsable_i18n_span(
+        "unparsableunicodesuggestion", character, lint, suggestion
+      )
+    )
+  }
+
+  # Check if code contains any other non-ASCII characters ----------------------
+  # Regex searches for any codepoints not in the ASCII range (00-7F)
+  non_ascii_pattern <- "[^\u01-\u7f]"
+  character <- str_extract(code, non_ascii_pattern)
+  lint <- exercise_highlight_unparsable_unicode(code, non_ascii_pattern)
+
+  return(build_unparsable_i18n_span("unparsableunicode", character, lint))
+}
+
+build_unparsable_i18n_span <- function(
+  key, character, lint, suggestion = NULL
+) {
```

Comment:
```suggestion
build_unparsable_i18n_span <- function(key, character, lint, suggestion = NULL) {
```

same again (all on one line or all on their own line)